app/objects/c_agent.py

```diff
@@ -127,19 +127,36 @@ async def calculate_sleep(self):
         return self.jitter('%d/%d' % (self.sleep_min, self.sleep_max))
 
     async def capabilities(self, ability_set):
+        """Get abilities that the agent is capable of running"""
         abilities = []
-        if self.executors:
-            preferred = self._get_preferred_executor()
-            executors = self.executors
-            for ai in set([pa.ability_id for pa in ability_set]):
-                total_ability = [ab for ab in ability_set if (ab.ability_id == ai)
-                                 and (ab.platform == self.platform) and (ab.executor in executors)]
-                if len(total_ability) > 0:
-                    val = next((ta for ta in total_ability if ta.executor == preferred), total_ability[0])
-                    if self.privileged_to_run(val):
-                        abilities.append(val)
+        for ability in ability_set:
+            if self.privileged_to_run(ability) and ability.get_executors(self.platform, self.executors):
+                abilities.append(ability)
         return abilities
 
+    async def capabilities_with_preference(self, ability_set):
+        """Get abilities that the agent is capable of running along with preferred executor"""
+        if not self.executors:
+            return []
+
+        ability_executors = []
+
+        for ability in ability_set:
+            if not self.privileged_to_run(ability):
+                continue
+            executor = await self.get_preferred_executor(ability)
+            if executor:
+                ability_executors.append((ability, executor))
+        return ability_executors
+
+    async def get_preferred_executor(self, ability):
+        """Get preferred executor for ability"""
```

Comment:
Can we document what the return values are?

Reply:
Documented a few agent methods.